libkbfs/impatient_debug_dumper.go

```diff
@@ -0,0 +1,162 @@
+package libkbfs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"runtime/pprof"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/keybase/client/go/logger"
+)
+
+type ctxTimeTracker struct {
+	ctx       context.Context
+	expiresAt time.Time
+
+	done int32
+}
+
+type ctxTimeTrackerListNode struct {
+	tracker *ctxTimeTracker
+	next    *ctxTimeTrackerListNode
+}
+
+type ctxTimeTrackerList struct {
+	front *ctxTimeTrackerListNode
+	tail  *ctxTimeTrackerListNode
+}
+
+func (c *ctxTimeTrackerList) peekFront() (t *ctxTimeTracker) {
+	if c.front == nil {
+		return nil
+	}
+	return c.front.tracker
+}
+
+func (c *ctxTimeTrackerList) popFront() (t *ctxTimeTracker) {
+	if c.front == nil {
+		return nil
+	}
+	t = c.front.tracker
+	c.front = c.front.next
+	if c.front == nil {
+		// last one!
+		c.tail = nil
+	}
+	return t
+}
+
+func (c *ctxTimeTrackerList) append(t *ctxTimeTracker) {
+	newNode := &ctxTimeTrackerListNode{
+		tracker: t,
+	}
+	if c.tail != nil {
+		c.tail.next = newNode
+	}
+	c.tail = newNode
+	if c.front == nil {
+		c.front = newNode
+	}
+}
+
+type ImpatientDebugDumper struct {
+	config Config
+	logger logger.Logger
+	dumpIn time.Duration
+
+	ticker  *time.Ticker
+	limiter *rate.Limiter
+
+	lock          sync.Mutex
+	chronological *ctxTimeTrackerList
+}
+
+const impatientDebugDumperCheckInterval = time.Second
+const impatientDebugDumperDumpMinInterval = time.Second // 1 dump per sec max
```

Comment:
Maybe `time.Minute`?

Reply:
I was worried 1m is too long for people stuck somewhere to wait for. How about 5s?

Reply:
I was talking about the rate-limiting, not the checking.  So there will still be one dump within a second of the expiration.  Isn't that good enough?  Why would we ever need one dump _per second_?

Reply:
Yes rate-limiting. I was thinking about the first ever dump which could be dropped if user restarts the app. This isn't one per second though; it's at most one per second. But I suppose I can change it into a ticker so there's always a tick there for the first dump even if we use 1m interval.

Reply:
Oh actually `rate.Limiter` works like that too. Nvm; will change!